### PR TITLE
Integration testing suite for CLI

### PR DESCRIPTION
**Pull request by thymikee**
**Summary**

This is a work in progress PR, where I'll be adding integration tests to Haul CLI. 
Will send followup PRs with more tests (unit included) after this is done.

TODOs:
- [x] setup framework for integration testing
- [x] add basic integration tests for commands

### CODE REVIEW

**Comment by thymikee** [comment]
I cannot yet figure out what's wrong with yarn install inside fixtures. It works locally. Also similar setup works fine inside Jest repo (also on circle 2, there's PR open).

Cc @bestander, maybe you have some smart thoughts?

**Comment by bestander** [comment]
@thymikee, this does ring a bell.

1. Try latest Yarn 0.28.4, it has heaps of bugs fixed

2. Any chance that you run `yarn install` in multiple threads?
If so use `--mutex network` argument to make sure installs don't break global cache

**Comment by thymikee** [comment]
@bestander thanks heaps! The `--mutex network` works just right 👌. I thought tests run serially, but I ran `npm run jest -i` instead of `npm run jest -- -i` . 
Thanks to `--mutex` I can investigate how the tests perform in parallel on Circle

**Comment by thymikee** [comment]
@grabbou I decided to skip testing Node 4 for now, as it's super counterproductive for me because of the Haul buggy behaviour there. Can we, for now, stop requiring `test-node-4`?

**Comment by thymikee** [comment]
There's still a lot of things to test, but at least we have a base now. I'll start adding some smaller unit ones in a followup to this

File: integration_tests/runHaul.js [hunk 1]
```diff
@@ -0,0 +1,65 @@
+/**
```
Comment (thymikee) [inline integration_tests/runHaul.js]: This file and files below are relevant

File: integration_tests/__tests__/start.test.js [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+      expect(stdout).toMatch('Haul is now bundling your React Native app');
+      expect(stdout).toMatch('A fresh build may take longer than usual');
+      expect(stdout).toMatch(
+        `You can now run the app on your ${platform} device`,
+      );
+      done();
+      haul.kill();
+    }
+  });
+}
```
Comment (thymikee) [inline integration_tests/__tests__/start.test.js]: This file and files above are relevant

**Comment by grabbou** [comment]
I think we can disable `node 4` tests for now - let's make an issue so that we can keep track of supporting them again.

File: integration_tests/__tests__/init.test.js [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+  haul.stdout.on('data', () => {
+    haul.stdin.write(ENTER_KEY);
+  });
+
+  haul.stdout.on('close', () => {
+    try {
+      const haulConfig = fs.readFileSync(CONFIG_FILE_PATH, 'utf8');
+      expect(haulConfig).toEqual(
+        // eslint-disable-next-line
+        'module.exports = ({ platform }) => ({\n  entry: `./index.${platform}.js`,\n});',
```
Comment (grabbou) [inline integration_tests/__tests__/init.test.js]: How about using snapshot for this?

File: integration_tests/__tests__/start.test.js [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+const stripAnsi = require('strip-ansi');
+
+const TEMP_DIR = path.resolve(os.tmpdir(), 'start_test');
+const TEST_PROJECT_DIR = path.resolve(
+  __dirname,
+  '../fixtures/react-native-with-haul',
+);
+
+beforeEach(() => {
+  run('yarn --mutex network', TEST_PROJECT_DIR);
```
Comment (grabbou) [inline integration_tests/__tests__/start.test.js]: Why do we want to run `yarn` before each of the suites? `start` doesn't affect `node_modules` at all?

Comment (thymikee) [inline integration_tests/__tests__/start.test.js]: I should switch it to beforeEach. You sure about that? I remember having troubles running it without react native available, but may revisit that

**Comment by grabbou** [comment]
Overall it looks good - does anyone else have any ideas / objections / suggestions for this pull request? 

@thymikee I think it's best if we pair on Slack for few seconds and I'll test what we got here before hitting final merge. What do you think?

File: integration_tests/__tests__/start.test.js (hunk 4 above)
Comment (grabbou) [inline integration_tests/__tests__/start.test.js]: That would be weird, but worth double checking

Comment (thymikee) [inline integration_tests/__tests__/start.test.js]: I mean beforeAll

**Comment by thymikee** [comment]
Addressed the feedback.

File: integration_tests/__tests__/bundle.test.js [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+const fs = require('fs');
+const { cleanup, run } = require('../utils');
+
+const TEST_PROJECT_DIR = path.resolve(
+  __dirname,
+  '../fixtures/react-native-with-haul',
+);
+
+beforeEach(() => {
+  run('yarn --mutex network', TEST_PROJECT_DIR);
```
Comment (grabbou) [inline integration_tests/__tests__/bundle.test.js]: Here I think it should be `beforeAll` as well

File: integration_tests/__tests__/start.test.js [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+    if (message.match(/ERROR/g)) {
+      done.fail(message);
+      haul.kill();
+    }
+
+    if (message.match('Built successfully!')) {
+      const stdout = messageBuffer.join('\n');
+      expect(stdout).toMatch('INFO  Ready at http://localhost:8081');
+      expect(stdout).toMatch('Haul is now bundling your React Native app');
+      expect(stdout).toMatch('A fresh build may take longer than usual');
```
Comment (grabbou) [inline integration_tests/__tests__/start.test.js]: Isn't it better to use `toMatchSnapshot` here as well? (just like we do for start message)

Comment (thymikee) [inline integration_tests/__tests__/start.test.js]: It is, but Android command produces some kind of warning with different wording on different machines, so I'd like to keep it this way for now. I can revisit later though

**Comment by thymikee** [comment]
Feedback addressed

**Pull request merged**